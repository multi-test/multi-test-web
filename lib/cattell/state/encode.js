import { BitStream } from 'bit-buffer';
import { toBase64 } from "./base64";
function answerToBit2(answer) {
    switch (answer) {
        case '': return 0b00;
        case 'A': return 0b01;
        case 'B': return 0b10;
        case 'C': return 0b11;
    }
}
function genderToBit2(gender) {
    switch (gender) {
        case '': return 0b00;
        case 'M': return 0b01;
        case 'F': return 0b10;
    }
}
export function encodeState(state) {
    const bitStream = new BitStream(new ArrayBuffer(49));
    {
        const byte = state.position;
        bitStream.writeUint8(byte);
    }
    for (const answer of state.answers) {
        const bit2 = answerToBit2(answer);
        bitStream.writeBits(bit2, 2);
    }
    {
        const bit2 = genderToBit2(state.profile.gender);
        bitStream.writeBits(bit2, 2);
    }
    {
        const bit6 = state.profile.age === 0 ? 0 : state.profile.age - 15;
        bitStream.writeBits(bit6, 6);
    }
    bitStream.index = 0;
    let base64Data = '';
    for (let i = 0; i < 65; i++) {
        base64Data += toBase64(bitStream.readBits(6, false));
    }
    const base64Name = btoa(unescape(encodeURIComponent(state.profile.name)));
    return base64Data + base64Name;
}
