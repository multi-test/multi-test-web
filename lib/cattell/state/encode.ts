import {BitStream} from 'bit-buffer';
import {CattellState} from "./type";
import {encode as encodeKOI8U} from '../../utils/koi8-u';
import {fromByteArray} from "../../utils/base64";
import {crc16} from "../../utils/crc16";

function answerToBit2(answer: '' | 'A' | 'B' | 'C'): 0 | 1 | 2 | 3 {
    switch (answer) {
        case '':  return 0b00;
        case 'A': return 0b01;
        case 'B': return 0b10;
        case 'C': return 0b11;
    }
}

function genderToBit2(gender: 'M' | 'F' | ''): 0 | 1 | 2 {
    switch (gender) {
        case '':  return 0b00;
        case 'M': return 0b01;
        case 'F': return 0b10;
    }
}

function ageToBit6(age: number /* 0, 16-70 */): number {
    return age === 0 ? 0 : age - 15;
}

export function encodeState(state: CattellState): string {
    const { position, answers, profile: { gender, age, name } } = state;

    const sizeInBytes = 49 + name.length + 2;
    const bitStream = new BitStream(new ArrayBuffer(sizeInBytes));
    bitStream.writeBits(0, 2);
    bitStream.writeUint8(position);

    for (const answer of answers) {
        bitStream.writeBits(answerToBit2(answer), 2);
    }

    bitStream.writeBits(genderToBit2(gender), 2);
    bitStream.writeBits(ageToBit6(age), 6);

    for (const byte of encodeKOI8U(name)) {
        bitStream.writeUint8(byte);
    }

    bitStream.index = 0;
    const body = bitStream.readArrayBuffer(sizeInBytes - 2);
    bitStream.writeUint16(crc16(body));

    bitStream.index = 0;
    const binary = bitStream.readArrayBuffer(sizeInBytes);
    return fromByteArray(binary);
}