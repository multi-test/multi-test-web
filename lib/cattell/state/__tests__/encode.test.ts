import {encodeState} from '../encode';
import createBlankCattellState from '../blank';
import {CattellState} from "../type";

describe('cattell.state.encode', () => {
    let state: CattellState;

    beforeEach(() => {
        state = createBlankCattellState();
    });

    afterEach(() => {
        const base64 = encodeState(state);
        expect(base64).toMatchSnapshot();
    });

    it('should create blank valid state matching the snapshot', () => {});

    it('should encode position', () => {
        state.position = 1;
    });

    it('should encode age', () => {
        state.profile.age = 70;
    });

    it('should encode name', () => {
        state.profile.name = 'Иванов '.repeat(10);
    });

    it('should encode male gender', () => {
        state.profile.gender = 'M';
    });

    it('should encode female gender', () => {
        state.profile.gender = 'F';
    });

    it('should encode A answers', () => {
        state.answers.fill('A');
    });

    it('should encode B answers', () => {
        state.answers.fill('B');
    });

    it('should encode C answers', () => {
        state.answers.fill('C');
    });

    it('should encode complete state', () => {
        for (let i = 0; i < 187; i++) {
            state.answers[i] = (i % 2 === 0) ? 'A' : (i % 3 === 0 ? 'B' : 'C');
        }

        state.position = 189;
        state.profile.gender = 'M';
        state.profile.age = 70;
        state.profile.name = 'Иванов И.И.';
    });
});

