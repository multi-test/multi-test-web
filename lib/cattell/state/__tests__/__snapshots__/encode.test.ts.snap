// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cattell.state.encode should create blank valid state matching the snapshot 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAJKn"`;

exports[`cattell.state.encode should encode A answers 1`] = `"AFRVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVAIjE"`;

exports[`cattell.state.encode should encode B answers 1`] = `"AKiqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqAKZh"`;

exports[`cattell.state.encode should encode C answers 1`] = `"APz/////////////////////////////////////////////////////////////ALwC"`;

exports[`cattell.state.encode should encode age 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA3GOt"`;

exports[`cattell.state.encode should encode complete state 1`] = `"9HZ2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d23enXwc7P1yDpLukuYpk="`;

exports[`cattell.state.encode should encode female gender 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAtCH"`;

exports[`cattell.state.encode should encode male gender 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAbO3"`;

exports[`cattell.state.encode should encode name 1`] = `"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOnXwc7P1yDp18HOz9cg6dfBzs/XIOnXwc7P1yDp18HOz9cg6dfBzs/XIOnXwc7P1yDp18HOz9cg6dfBzs/XIOnXwc7P1yBUhQ=="`;

exports[`cattell.state.encode should encode position 1`] = `"BAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEBy"`;
